File: frontend/src/types/inventory.ts
export interface InventoryItem {
  id: number; // Represents the Product ID
  product_name: string;
  sku: string;
  current_quantity: number;
  minimum_quantity: number;
  maximum_quantity: number;
  status: string; // "Low Stock" or "In Stock"
  is_low_stock: boolean;
}

export interface StockMovement {
  id: number;
  product_id: number;
  product_name: string;
  sku: string;
  movement_type: "IN" | "OUT";
  quantity: number;
  remarks: string | null;
  created_by: number;
  created_at: string;
}

export interface StockInRequest {
  product_id: number;
  quantity: number;
  remarks?: string;
}

export interface StockOutRequest {
  product_id: number;
  quantity: number;
  remarks?: string;
}

export interface MovementFilters {
  product_id?: number;
  type?: "IN" | "OUT" | "";
  from_date?: string;
  to_date?: string;
}

File: backend/app/schemas/inventory.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict, model_validator


class StockInRequest(BaseModel):
    product_id: int
    quantity: int = Field(..., gt=0, description="Quantity to stock in must be > 0")
    remarks: Optional[str] = Field(None, max_length=255)


class StockOutRequest(BaseModel):
    product_id: int
    quantity: int = Field(..., gt=0, description="Quantity to stock out must be > 0")
    remarks: Optional[str] = Field(None, max_length=255)


class StockMovementOut(BaseModel):
    id: int
    product_id: int
    product_name: str
    sku: str
    movement_type: str
    quantity: int
    remarks: Optional[str] = None
    created_by: int
    created_at: datetime

    class Config:
        from_attributes = True


class InventoryOut(BaseModel):
    id: int  # Product ID
    product_id: int
    product_name: str
    sku: str
    current_quantity: int = Field(..., ge=0)
    minimum_quantity: int = Field(..., ge=0)
    maximum_quantity: int = Field(..., ge=0)
    status: str  # "Low Stock" vs "In Stock"
    is_low_stock: bool
    updated_at: datetime

    @model_validator(mode='after')
    def compute_low_stock(self) -> "InventoryOut":
        self.is_low_stock = self.current_quantity <= self.minimum_quantity
        return self

    model_config = ConfigDict(from_attributes=True)
